### PR TITLE
Support backprop in `default.mixed` device for autograd

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Currently only the parameter-shift differentiation method is supported in the `default.mixed` device. To enable better integration with auto-differentiation frameworks like autograd, it would be good to support backpropagation in this device.

**Description of the Change:**
Now backpropagation is supported in the `default.mixed` device for the autograd interface. (Other interfaces to come in the future.)

Note that with how QNodes choose the best differentiation method, all QNodes with the `default.mixed` device will now default to using `diff_method=backprop` when using the autograd interface.

**Benefits:**
Better integration with auto-differentiation frameworks.

**Possible Drawbacks:**
Might break some workflows dependent on using parameter-shift as default gradient method for `default.mixed`.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/1528

### CODE REVIEW

File: tests/devices/test_default_mixed_autograd.py [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliY(0))
+
+        expected = -np.sin(p)
+
+        assert circuit.gradient_fn == "backprop"
+        assert np.isclose(circuit(p), expected, atol=tol, rtol=0)
+
+    def test_correct_state(self, tol):
+        """Test that the device state is correct after applying a
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:88]: minor
```suggestion
        """Test that the device state is correct after evaluating a
```

File: tests/devices/test_default_mixed_autograd.py [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+        "measurement",
+        [
+            qml.expval(qml.PauliY(0)),
+            qml.var(qml.PauliY(0)),
+            qml.probs(wires=[1]),
+            qml.probs(wires=[2, 0]),
+        ],
+    )
+    def test_real_dtype(self, r_dtype, measurement, tol):
+        """Test that the default qubit plugin provides correct result for a simple circuit"""
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:131]: This docstring seems to describe a different test case.

File: tests/devices/test_default_mixed_autograd.py [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+    @pytest.mark.parametrize(
+        "measurement",
+        [qml.state(), qml.density_matrix(wires=[1]), qml.density_matrix(wires=[2, 0])],
+    )
+    def test_complex_dtype(self, c_dtype, measurement, tol):
+        """Test that the default qubit plugin provides correct result for a simple circuit"""
+        p = 0.543
+
+        dev = qml.device("default.mixed", wires=3, c_dtype=c_dtype)
+        # dev.C_DTYPE = c_dtype
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:155]:
```suggestion
```

File: tests/devices/test_default_mixed_autograd.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+            return qml.expval(qml.PauliZ(0))
+
+        a = np.array(0.54, requires_grad=True)
+
+        def cost(a):
+            """A function of the device quantum state, as a function
+            of input QNode parameters."""
+            circuit(a)
+            res = np.abs(dev.state) ** 2
+            return res[1][1] - res[0][0]
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:280]: This is a great test case, though maybe it's best to return `qml.state()` from the QNode directly. That pipeline should use `dev.state` too under the hood.

Would also be cool to test a QNode that returns `qml.density_matrix(wires=wires)` for some `wires`.

File: tests/devices/test_default_mixed_autograd.py [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+        """Tests that an error is raised if diff_method='backprop' but not using
+        the Autograd interface"""
+        dev = qml.device("default.mixed", wires=1)
+
+        def circuit(x, w=None):
+            qml.RZ(x, wires=w)
+            return qml.expval(qml.PauliX(w))
+
+        msg = "Device default.mixed only supports diff_method='backprop' when using the autograd interface"
+        msg = re.escape(msg)
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:418]: Oh, this is great, haven't seen it being used before!

File: tests/devices/test_default_mixed_autograd.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+        expected_cost = 0.5 * (np.cos(a) * np.cos(b) + np.cos(a) - np.cos(b) + 1)
+        assert np.allclose(res, expected_cost, atol=tol, rtol=0)
+
+        res = qml.grad(circuit)(a, b)
+        expected_grad = np.array(
+            [-0.5 * np.sin(a) * (np.cos(b) + 1), 0.5 * np.sin(b) * (1 - np.cos(a))]
+        )
+        assert np.allclose(res, expected_grad, atol=tol, rtol=0)
+
+    @pytest.mark.skip
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:335]: How come this is being skipped?

File: tests/devices/test_default_mixed_autograd.py [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+        )
+        assert np.allclose(res[0], expected[0], atol=tol, rtol=0)
+        assert np.allclose(res[1], expected[1], atol=tol, rtol=0)
+
+
+@pytest.mark.autograd
+class TestHighLevelIntegration:
+    """Tests for integration with higher level components of PennyLane."""
+
+    @pytest.mark.skip
```
Comment (antalszava) [inline tests/devices/test_default_mixed_autograd.py:468]: How come this is being skipped?

**Review by antalszava — COMMENTED** [review]
This is looking amazing @eddddddy! :fireworks: 

The addition looks great overall, would just be keen to check answers to the comments made.

Perhaps the following test case could also be added:
```py
class TestSample:
    """Tests for the sample integration"""

    def test_backprop_error(self):
        """Test that sampling in backpropagation mode raises an error"""
        dev = qml.device("default.mixed", wires=2)

        @qnode(dev, diff_method="backprop")
        def circuit():
            qml.RX(0.54, wires=0)
            return qml.sample(qml.PauliZ(0)), qml.sample(qml.PauliX(1))

        with pytest.raises(qml.QuantumFunctionError, match="only supported when shots=None"):
            circuit(shots=10)
```

File: pennylane/devices/default_mixed_autograd.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/devices/default_mixed_autograd.py]: I removed this file a while ago. The `default.mixed` device is now used for all interfaces.

File: tests/devices/test_default_mixed.py [hunk 9]
```diff
@@ -923,6 +923,20 @@ def test_apply_state_vector_wires(self, tol):
 
         assert np.allclose(dev.state, root_state(nr_wires), atol=tol, rtol=0)
 
+    def test_apply_state_vector_subsystem(self, tol):
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/devices/test_default_mixed.py:926]: There's no bug. It was just added for coverage that was previously missing (lines 394-410 in `devices/default_mixed.py`):
```python
else:
    # generate basis states on subset of qubits via the cartesian product
    basis_states = qnp.asarray(
        list(itertools.product([0, 1], repeat=len(device_wires))), dtype=self.C_DTYPE
    )

    # get basis states to alter on full set of qubits
    unravelled_indices = qnp.zeros((2 ** len(device_wires), self.num_wires), dtype=int)
    unravelled_indices[:, device_wires] = basis_states

    # get indices for which the state is changed to input state vector elements
    ravelled_indices = qnp.ravel_multi_index(unravelled_indices.T, [2] * self.num_wires)

    state = qnp.scatter(ravelled_indices, state, [2**self.num_wires])
    rho = qnp.outer(state, qnp.conj(state))
    rho = qnp.reshape(rho, [2] * 2 * self.num_wires)
    self._state = qnp.asarray(rho, dtype=self.C_DTYPE)
```

File: pennylane/devices/default_mixed.py [hunk 10]
```diff
@@ -133,16 +153,16 @@ def _create_basis_state(self, index):
...
-        return self._reshape(rho, [2] * (2 * self.num_wires))
+        rho = qnp.asarray(rho, dtype=self.C_DTYPE)
+        return qnp.reshape(rho, [2] * (2 * self.num_wires))
 
     @classmethod
     def capabilities(cls):
         capabilities = super().capabilities().copy()
         capabilities.update(
-            returns_state=True,
+            returns_state=True, supports_inverse_operations=True, passthru_interface="autograd"
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/devices/default_mixed.py:165]: I was trying to match the capabilities of `default.qubit` as closely as possible. I haven't extensively tested that inverse operations work so I'll remove it for this PR.

File: tests/devices/test_default_mixed_autograd.py (hunk 6 above)
Comment (eddddddy) [reply tests/devices/test_default_mixed_autograd.py:334]: Oh, I forgot to add it back in after our previous discussion on handling this.

File: tests/devices/test_default_mixed_autograd.py (hunk 7 above)
Comment (eddddddy) [reply tests/devices/test_default_mixed_autograd.py:467]: The test is for whether the Hamiltonian is being split into its different terms when computing `expval`. This also seems out of scope of this PR, so I'll just remove it.

File: pennylane/devices/default_mixed.py [hunk 11]
```diff
@@ -473,14 +499,15 @@ def _apply_operation(self, operation):
...
-                density_matrix = self._reshape(self._state, (dim, dim))
+                density_matrix = qnp.reshape(self._state, (dim, dim))
                 if operation.tag:
                     self._debugger.snapshots[operation.tag] = density_matrix
                 else:
                     self._debugger.snapshots[len(self._debugger.snapshots)] = density_matrix
             return
 
         matrices = self._get_kraus(operation)
+        matrices = [qnp.cast(mat, dtype=self.C_DTYPE) for mat in matrices]
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/devices/default_mixed.py:510]: Yes, I was thinking the dtype of the state might be coerced to a larger dtype (such as `complex128` when the user-provided dtype is `complex64`). However, I see we already do this cast in `_apply_channel` anyways, so this line isn't necessary.

[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/devices/default_mixed.py]: Removed the unnecessary cast.

File: doc/releases/changelog-dev.md [hunk 12]
```diff
@@ -180,6 +180,31 @@
...
+  [(#2615)](https://github.com/PennyLaneAI/pennylane/pull/2615)
+
+  As a result, the default differentiation method for the device is now `"backprop"`. To continue using the old default `"parameter-shift"`, explicitly specify this differentiation method in the QNode.
+
+  **Example**
+
+  ```python
+  dev = qml.device("default.mixed", wires=2)
+
+  @qml.qnode(dev, interface="autograd")
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply doc/releases/changelog-dev.md]: Added `diff_method="backprop"` in changelog entry

File: pennylane/devices/default_mixed.py [hunk 13]
```diff
@@ -228,27 +248,29 @@ def _apply_channel(self, kraus, wires):
...
+
+        kraus = qnp.stack(kraus)
+        kraus_dagger = qnp.stack(kraus_dagger)
 
         # Changes tensor shape
         if kraus[0].shape[0] == kraus[0].shape[1]:
             kraus_shape = [len(kraus)] + [2] * num_ch_wires * 2
-            kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
-            kraus_dagger = self._cast(self._reshape(kraus_dagger, kraus_shape), dtype=self.C_DTYPE)
+            kraus = qnp.reshape(qnp.cast(kraus, dtype=self.C_DTYPE), kraus_shape)
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/devices/default_mixed.py]: Changed to old order.

File: tests/devices/test_default_mixed_autograd.py (hunk 2 above)
Comment (eddddddy) [reply tests/devices/test_default_mixed_autograd.py]: Changed the docstring to match what is tested.

File: tests/devices/test_default_mixed_autograd.py (hunk 4 above)
Comment (eddddddy) [reply tests/devices/test_default_mixed_autograd.py]: Changed the QNode to directly return the state, and also added another test for `qml.density_matrix`.

File: tests/devices/test_default_mixed_autograd.py [hunk 14]
```diff
@@ -0,0 +1,585 @@
...
+        def cost(weights):
+            return np.sum(qnodes(weights))
+
+        grad = qml.grad(cost)(weights)
+        assert grad.shape == weights.shape
+
+
+@pytest.mark.autograd
+class TestOps:
+    """Unit tests for operations supported by the default.qubit.autograd device"""
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/devices/test_default_mixed_autograd.py]: Moved the op tests before the integration tests.

File: pennylane/devices/default_mixed.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/devices/default_mixed.py]: Removed the device capability.

File: pennylane/ops/channel.py [hunk 15]
```diff
@@ -81,8 +82,12 @@ def compute_kraus_matrices(gamma):  # pylint:disable=arguments-differ
         if not 0.0 <= gamma <= 1.0:
             raise ValueError("gamma must be in the interval [0,1].")
 
-        K0 = np.diag([1, np.sqrt(1 - gamma)])
-        K1 = np.sqrt(gamma) * np.array([[0, 1], [0, 0]])
+        # small number to add to probabilities to prevent dividing by
+        # zero in the gradient of sqrt
+        eps = 1e-14
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/ops/channel.py]: I changed it so that:
1. `qml.math` now has an attribute `eps`.
2. All channel ops now use `qml.math.eps` instead of defining their own numerical epsilon.
3. The user can modify this by setting e.g. `qml.math.eps = 1e-8`.

Not sure if this is the cleanest solution so would love to know your thoughts.

[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/ops/channel.py]: I think that sounds great! :)

**Review by antalszava — APPROVED** [review]
Looks good! :tada: :fireworks: This is an awesome addition :blush:

File: tests/devices/test_default_mixed_autograd.py (hunk 7 above)
Comment (eddddddy) [reply tests/devices/test_default_mixed_autograd.py]: Removed this test.

**Comment by antalszava** [comment]
[sc-21986]
